### PR TITLE
networkd: request a address dump to update the flags of addresses.

### LINKED ISSUE #9130: IPv6 point-to-point link stuck in state "configuring"
**systemd version the issue has been seen with**
237 (237-3~bpo9+1)

**Used distribution**
Debian Stretch

**Expected behaviour you didn't see**
Interface state configured

**Unexpected behaviour you saw**
Interface state remains in state "configuring"

**Steps to reproduce the problem**

Add a simple network configuration file `/etc/systemd/network/eth0.network`:

```
[Match]
Name=eth0

[Address]
Address=2001:db8::20
Peer=2001:db8::10/128
```

Then restart systemd-networkd.

The interface is apparently configured, and the link works, but systemd-networkd shows its state as "routable (configuring)":

```
$ sudo networkctl status eth0
● 2: eth0
       Link File: /lib/systemd/network/99-default.link
    Network File: /etc/systemd/network/eth0.network
            Type: ether
           State: routable (configuring)
            Path: acpi-VMBUS:01
          Driver: hv_netvsc
      HW Address: 00:1d:d8:b7:86:9c (Microsoft Corporation)
         Address: 2001:db8::20
                  fe80::21d:d8ff:feb7:869c
```

The link works; I can ping and SSH across to the peer.

```
$ ip address show dev eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 00:1d:d8:b7:86:9c brd ff:ff:ff:ff:ff:ff
    inet6 2001:db8::20 peer 2001:db8::10/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::21d:d8ff:feb7:869c/64 scope link 
       valid_lft forever preferred_lft forever
```

```
$ ip -6 route
2001:db8::10 dev eth0 proto kernel metric 256  pref medium
2001:db8::20 dev eth0 proto kernel metric 256  pref medium
fe80::/64 dev eth0 proto kernel metric 256  pref medium
```

The log from networkd is available here:
[networkd-log.txt](https://github.com/systemd/systemd/files/2052825/networkd-log.txt)

I noticed that it was trying to send some router solicitations:

> NDISC: Started IPv6 Router Solicitation client
> NDISC: Sent Router Solicitation, next solicitation in 3s
> NDISC: Sent Router Solicitation, next solicitation in 8s

Thinking that this may be the problem, I tried disabling `IPv6AcceptRA`:

```
[Match]
Name=eth0

[Network]
IPv6AcceptRA=no

[Address]
Address=2001:db8::20
Peer=2001:db8::10/128
```

However, the interface remains in state configuring. The log file is available here:
[networkd-log-no-ra.txt](https://github.com/systemd/systemd/files/2052849/networkd-log-no-ra.txt)

### PR DESCRIPTION
**Pull request by ssahani**
1. Request a address dump.
2. When we do a dump and peer address is configured the peer is
given by kernel . Hence add the peer when it is available

Closes #9130

### CODE REVIEW

**Comment by poettering** [comment]
@ssahani hmm, so is this going to fix #8686, #5043, #7315, #5154 too? (and possibly more?

**Comment by ssahani** [comment]
#8686, #5043, #7315 yes not tested with #5154 But seems it will do.

**Comment by poettering** [comment]
@ssahani any chance you can rework this PR like that? would love to merge a fix for this before v239

**Comment by ssahani** [comment]
https://elixir.bootlin.com/linux/v4.1.4/source/net/ipv6/addrconf.c#L880 The kernel actaully assigns the flag by default. Then it performs DAD if that fails then it updates. I am not quite sure why we have to assign the flag

**Comment by keszybz** [comment]
So what's the status here? Does it need a rework or not? @teg, can you pitch in?

**Comment by ssahani** [comment]
I figure we can query the kernel to know the address state / flag . After adding a address how much time we need to wait to ask kernel for the address state ?

This is what I think.

1. Add a address make it tentative and put it on a list with a priority queue.
2.  walk this PQ when timer expired and query the kernel. Would this be the design approach we follow ? 

@poettering @yuwata @poettering  ?

**Comment by ssahani** [comment]
@poettering @yuwata PTAL.

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -962,3 +966,37 @@ bool address_is_ready(const Address *a) {
         else
                 return !(a->flags & (IFA_F_TENTATIVE | IFA_F_DEPRECATED));
 }
+
+
```
Comment (yuwata) [inline src/network/networkd-address.c]: Doubled empty lines.

File: src/network/networkd-link.h [hunk 2]
```diff
@@ -125,6 +125,8 @@ typedef struct Link {
         unsigned lldp_tx_fast; /* The LLDP txFast counter (See 802.1ab-2009, section 9.2.5.18) */
         sd_event_source *lldp_emit_event_source;
 
+        sd_event_source *address_event_source;
```
Comment (yuwata) [inline src/network/networkd-link.h:128]: This seems not unref()ed. Please add `sd_event_source_unref()` at `link_free()`.

**Comment by ssahani** [comment]
updated thanks @yuwata

**Comment by yuwata** [comment]
@ssahani I've tried to add a testcase for this at https://github.com/yuwata/systemd/tree/pr-9174-follow-ups.
Then I found that this does not fix #9130, that is, `networkctl` still shows the link status 'configuring':
```
sudo ./systemd-networkd-tests.py NetworkdNetWorkTests.test_address_peer
modprobe: FATAL: Module vcan not found in directory /lib/modules/4.19.6-200.fc28.x86_64
modprobe: FATAL: Module wireguard not found in directory /lib/modules/4.19.6-200.fc28.x86_64
test_address_peer (__main__.NetworkdNetWorkTests) ... 
6: dummy98: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether fa:40:e0:6c:e6:d7 brd ff:ff:ff:ff:ff:ff
    inet 10.2.3.4 peer 10.2.3.5/16 scope global 32
       valid_lft forever preferred_lft forever
    inet 10.6.7.8/16 brd 10.6.255.255 scope global 33
       valid_lft forever preferred_lft forever
    inet6 fe80::f840:e0ff:fe6c:e6d7/64 scope link 
       valid_lft forever preferred_lft forever
● 6: dummy98
       Link File: /usr/lib/systemd/network/99-default.link
    Network File: /run/systemd/network/25-address-section.network
            Type: ether
           State: routable (configuring)
          Driver: dummy
      HW Address: fa:40:e0:6c:e6:d7
         Address: 10.2.3.4
                  10.6.7.8
                  fe80::f840:e0ff:fe6c:e6d7
FAIL

======================================================================
FAIL: test_address_peer (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 656, in test_address_peer
    self.assertRegex(output, 'State: routable (configured)')
AssertionError: Regex didn't match: 'State: routable (configured)' not found in '● 6: dummy98\n       Link File: /usr/lib/systemd/network/99-default.link\n    Network File: /run/systemd/network/25-address-section.network\n            Type: ether\n           State: routable (configuring)\n          Driver: dummy\n      HW Address: fa:40:e0:6c:e6:d7\n         Address: 10.2.3.4\n                  10.6.7.8\n                  fe80::f840:e0ff:fe6c:e6d7'

----------------------------------------------------------------------
Ran 1 test in 9.499s

FAILED (failures=1)
```

**Comment by yuwata** [comment]
So, I guess `link_check_ready()` needs to be called after enumerating address, though not tested.

**Comment by yuwata** [comment]
BTW, setting up timer and enumerating addresses for each link seems to be redundant for me. How about managing the timer by `Manager` and make its callback set timer again if there exist not-ready addresses?

**Comment by yuwata** [comment]
@ssahani I noticed that it is sufficient to check link is ready when address is updated at least for fixing #9130. Please see #11232.

**Comment by ssahani** [comment]
@yuwata we are trying to fix the DAD for IPV6 which kernel takes some time to finish. So we need to query to the kernel the address dump . when we add a ipv6 address it makes it tentative and starts the DAD. once successful it removes the flag and mark is good one.

**Comment by yuwata** [comment]
Ah, I've not tested for ipv6 cases... Thanks. But, at least for ipv4 `Address.Peer=` case, #11232 seems to be sufficient.

**Comment by ssahani** [comment]
yes we actually should update. looks good to me .

**Comment by ssahani** [comment]
@yuwata  Sorry for delay 

I rested again with the configuration

```
[root@Zeus network]# cat dummy1.network
[Match]
Name=dummy1

[Address]
Address=2001:db8::20
Peer=2001:db8::10/128

```

```
10: dummy1: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 1a:7f:6e:e2:7b:19 brd ff:ff:ff:ff:ff:ff
    inet6 2001:db8::20 peer 2001:db8::10/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::187f:6eff:fee2:7b19/64 scope link 
       valid_lft forever preferred_lft forever
[root@Zeus network]# networkctl 
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           carrier     unmanaged 
 10 dummy1           ether              routable    configured <==========================Here
```
It does fixes the issue

**Comment by yuwata** [comment]
I hope the revised version #11428 is sufficient to fix the issue.

**Pull request closed without merging**